packages/aws-cdk-lib/aws-stepfunctions/lib/states/distributed-map.ts

```diff
@@ -35,6 +36,15 @@ export interface DistributedMapProps extends MapBaseProps {
    */
   readonly itemReader?: IItemReader;
 
+  /**
+   * ItemReaderPath
+   *
+   * Configuration for where to read items dataset in S3 to iterate, as JsonPath
+   *
+   * @default - No itemReaderPath
+   */
+  readonly itemReaderPath?: IItemReaderPath;
```

Comment:
Following is intentional, with `IItemReaderPath`, during compile time, we will not have information about `bucket`, hence, we can't add any permission during compile time, therefore, this action needs to be taken separately by caller who starts `StateMachine` execution with `bucketName`. Also, if considered adding `*` policy statement, then it will violate minimum access to resource guideline.
> The main difference I see between the ItemReader and ItemReaderPath is for the first one, we automatically add the policy to access the S3 bucket, and the other one is we could not add this policy. 

Following is precisely issue with current codebase that it is not extensible to allow changes and most of such options are already considered to decent extent (refer PR description) - and among those, this approach was least troublesome.
> We need to think about a better design approach.
> one of the options is to deprecate the current ItemReader class, and add a new one, but I also do not like the idea of keep deprecating existing properties for new properties and ask customers to update their code.

Now, if you have other alternatives, feel free to pitch in. This is exactly why this PR is raised within minimum constructs.

Reply:
Similar duplicate parameters are also utilised in different places ([reference](https://github.com/ChakshuGupta13/aws-cdk/blob/issue_29409/packages/aws-cdk-lib/aws-stepfunctions/lib/states/fail.ts#L30)) - although they have simple types, not full-fledged classes (or interfaces)
```
  /**
   * Error code used to represent this failure
   *
   * @default - No error code
   */
  readonly error?: string;

  /**
   * JsonPath expression to select part of the state to be the error to this state.
   *
   * You can also use an intrinsic function that returns a string to specify this property.
   * The allowed functions include States.Format, States.JsonToString, States.ArrayGetItem, States.Base64Encode, States.Base64Decode, States.Hash, and States.UUID.
   *
   * @default - No error path
   */
  readonly errorPath?: string;
```
> I am finding this approach very confusing to customers. Also we are duplicating every thing in 2 versions of almost the same classes

Reply:
Original proposal required breaking changes but avoided code duplication, hence also asked in ticket:
![IItemReader-Proposal](https://github.com/user-attachments/assets/806a294d-aba0-40b6-a6f2-6e91d6e5b3a2)

Reply:
@moelasmar thanks for a second review and provide the feedback. I do agree that a number of resources are duplicate between `IItemReader` and `IIteamReaderPath`. This is not the best design but as mentioned earlier, we can't directly modify `IItemReader` due to breaking changes. Specifically `IItemReader` takes a required parameter `bucket: IBucket` and this doesn't work for dynamic parameter handling. 

I agree this is not the best practice but there are many places in CDK that has similar issues. Would like you get your thought on the next step. One thing I can think of is to create a base interface that defines the shared parameters and `IItemReader` and `IItemReaderPath` to extend the interface and has its own definitions from there.

Reply:
I tried to think about base interface as we discussed briefly (re:Slack & WorkLog) but then got stuck as to how will that be better useful? We will still need different implementations (both for static and dynamic) for each reader type (object, JSON, CSV, manifest etc.)
> One thing I can think of is to create a base interface that defines the shared parameters and IItemReader and IItemReaderPath to extend the interface and has its own definitions from there.

So, can you please share if you foresee any benefit with base class (and if possible, can you please explain "how" like an example - that will help clarify few thoughts for me).

---

Secondly, I was thinking about examples like following:
```
 /**
   * Error code used to represent this failure
   *
   * @default - No error code
   */
  readonly error?: string;

  /**
   * JsonPath expression to select part of the state to be the error to this state.
   *
   * You can also use an intrinsic function that returns a string to specify this property.
   * The allowed functions include States.Format, States.JsonToString, States.ArrayGetItem, States.Base64Encode, States.Base64Decode, States.Hash, and States.UUID.
   *
   * @default - No error path
   */
  readonly errorPath?: string;
```
In such example, we always have two parameters (one for static and one for dynamic), so, now if `IItemReader` and its implementations have X number of static parameters (e.g. `bucket`, `prefix`, `key` etc.), then to have dynamic alternatives, we will have same number of parameters (e.g. `bucketNamePath`, `prefixPath`, `keyPath` etc.).

Also:
- `render` method for both use-cases will differ (e.g. for dynamic `Parameters` will be suffixed with `.$`) 
- As mentioned earlier, for dynamic use-case, we cannot have `policyStatements` defined statically since required information is unavailable at that time, hence, that will also differ. 

Therefore, only thing common between static and dynamic alternatives that can be de-duplicated is `resource` which is initialised to an `Arn` in `constructor`, hence, by making base class (or trying to generalise these classes), we can only avoid duplication of `resource` at max (since, we need do parameters like `bucketNamePath`, `prefixPath`, `keyPath` etc.)

--- 

Ideal solution would be to have classes like following:

```
export interface IItemReader {
  /** bucketName and bucketNamePath mutually exclusive*/
  readonly bucketName?: string;
  readonly bucketNamePath?: string;

  readonly resource: string;
  readonly maxItems?: number;
  render(): any;
}

export class S3ObjectsItemReader implements IItemReader {
  readonly bucketName?: string;
  readonly bucketNamePath?: string;

  /** prefix and prefixPath mutually exclusive*/
  readonly prefix?: string;
  readonly prefixPath?: string;

  readonly resource: string;
  readonly maxItems?: number;

  constructor(props...) {
    this.bucketName = props.bucketName;
    this.bucketNamePath = props.bucketNamePath;
    this.prefix = props.prefix;
    this.prefixPath = props.prefixPath;
    this.maxItems = props.maxItems;
    this.resource = Arn.format({
      region: '',
      account: '',
      partition: Aws.PARTITION,
      service: 'states',
      resource: 's3',
      resourceName: 'listObjectsV2',
      arnFormat: ArnFormat.COLON_RESOURCE_NAME,
    });
  }

  public render(): any {
    return FieldUtils.renderObject({
      Resource: this.resource,
      ...(this.maxItems && { ReaderConfig: { MaxItems: this.maxItems } }),
      Parameters: {
        ... (this.bucketName && { Bucket: this.bucketName } ),
        ... (this.bucketNamePath && { 'Bucket.$': this.bucketNamePath } ),
        ...(this.prefix && { Prefix: this.prefix }),
        ...(this.prefixPath && { 'Prefix.$': this.prefixPath }),
      },
    });
  }
}
```
This way we can even have partial static parameters and rest dynamic - example:
```ts
const distributedMap = new sfn.DistributedMap(this, 'DistributedMap', {
    itemReaderPath: new sfn.S3ObjectsItemReader({
      bucketName: 'my-bucket',
      prefixPath: '$.prefix',
    }),
  });
  distributedMap.itemProcessor(new sfn.Pass(this, 'Pass'));
```
As much as I like this freedom to choose which parameter needs to be dynamic, this will need breaking changes.

Reply:
@moelasmar @GavinZZ 
my colleague discovered following way with which we can provide dynamic parameters (except for Bucket because it is currently of type `IBucket` instead of `string`):
```
const myBucket = new Bucket(stack, 'MyBucket', ...);
const distributedMap = new sfn.DistributedMap(this, 'DistributedMap', {
    itemReaderPath: new sfn.S3ObjectsItemReader({
      bucket: myBucket,
      prefix: JsonPath.stringAt('$.prefix;),
    }),
  });
  distributedMap.itemProcessor(new sfn.Pass(this, 'Pass'));
```
This code snippet produces following `cdk.out` template:
```
...
"Parameters\":{\"Bucket\":\"",
       {
        "Ref": "MyBucket..."
       },
       "\",\"Prefix.$\":\"$.prefix\"}}}}}"
      ]
...
```
Not sure why this works but since it works, this PR may not be needed for the said issue (to be able to pass dynamic file name).

---
Now, only one issue remains: how to also allow `Bucket` to be dynamic? For that, the ideal would be breaking change where `bucket: IBucket` is changed to `bucket: string`.

Reply:
Nice, but we still need to fix the issue where it only takes `IBucket`. I don't find much values actually to move the common parameters into a base interface and have `IItemReader` and `IItemReaderPath` to extend it.

@moelasmar would you mind take another look and provide some feedback if you have any ideas on a non-breaking change but don't have to duplicate the same properties.